lib/common/src/Amp.php

```diff
@@ -0,0 +1,261 @@
+     *
+     * @see https://github.com/ampproject/amphtml/blob/master/spec/amp-cache-guidelines.md#guidelines-adding-a-new-cache-to-the-amp-ecosystem
+     *
+     * @var array[]
+     */
+    const DYNAMIC_COMPONENTS = [
+        Attribute::CUSTOM_ELEMENT  => [Extension::GEO],
+        Attribute::CUSTOM_TEMPLATE => [],
+    ];
+
+    /**
+     * Array of custom element names that delay rendering.
+     *
+     * @var string[]
+     */
+    const RENDER_DELAYING_EXTENSIONS = [
+        Extension::DYNAMIC_CSS_CLASSES,
+        Extension::EXPERIMENT,
+        Extension::STORY,
+    ];
+
+    /**
+     * Standard boilerplate CSS stylesheet.
+     *
+     * @var string
+     */
+    const BOILERPLATE_CSS = 'body{-webkit-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-moz-animation:-amp-start 8s steps(1,end) 0s 1 normal both;-ms-animation:-amp-start 8s steps(1,end) 0s 1 normal both;animation:-amp-start 8s steps(1,end) 0s 1 normal both}@-webkit-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-moz-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-ms-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@-o-keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}@keyframes -amp-start{from{visibility:hidden}to{visibility:visible}}';
+
+    /**
+     * Boilerplate CSS stylesheet for the <noscript> tag.
+     *
+     * @var string
+     */
+    const BOILERPLATE_NOSCRIPT_CSS = 'body{-webkit-animation:none;-moz-animation:none;-ms-animation:none;animation:none}';
+
+    /**
+     * Boilerplate CSS stylesheet for Amp4Ads & Amp4Email.
+     *
+     * @var string
+     */
+    const AMP4ADS_AND_AMP4EMAIL_BOILERPLATE_CSS = 'body{visibility:hidden}';
+
+    /**
+     * Amp runtime tag name.
+     *
+     * @var string
+     */
+    const RUNTIME = 'amp-runtime';
+
+    // Amp classes reserved for internal use.
+    const LAYOUT_ATTRIBUTE          = 'i-amphtml-layout';
+    const NO_BOILERPLATE_ATTRIBUTE  = 'i-amphtml-no-boilerplate';
+    const LAYOUT_CLASS_PREFIX       = 'i-amphtml-layout-';
+    const LAYOUT_SIZE_DEFINED_CLASS = 'i-amphtml-layout-size-defined';
+    const SIZER_ELEMENT             = 'i-amphtml-sizer';
+    const INTRINSIC_SIZER_ELEMENT   = 'i-amphtml-intrinsic-sizer';
+
+    /**
+     * Check if a given node is the Amp runtime script.
+     *
+     * The Amp runtime script node is of the form '<script async src="https://cdn.ampproject.org...v0.js"></script>'.
+     *
+     * @param DOMNode $node Node to check.
+     * @return bool Whether the given node is the Amp runtime script.
+     */
+    public static function isRuntimeScript(DOMNode $node)
+    {
+        if (
+            ! self::isAsyncScript($node)
+            || self::isExtension($node)
+        ) {
+            return false;
+        }
+
+        $src = $node->getAttribute(Attribute::SRC);
+
+        if (strpos($src, self::CACHE_ROOT_URL) !== 0) {
```

Comment:
How about using `substr()` as is being used below? No need to search the entire string if it is supposed to start at.

Reply:
For potentially long strings, that would make sense. However, here, we're dealing with a URL.

Using `substr()` with a constant would mean we need to calculate the length of the constant first before doing the comparison, so I don't think it will be faster, and it will be less maintainable.

Reply:
I don't feel strongly about this. But using the same argument, why does the code below use `substr()`:

```php
            substr($src, -6) !== '/v0.js'
            && substr($src, -14) !== '/amp4ads-v0.js'
```

Reply:
That code doesn't need to calculate the length, it is a known value. In that case, `substr()` is faster.